pkg/services/datasources/service/datasource.go

```diff
@@ -534,11 +534,11 @@ func (s *Service) UpdateDataSource(ctx context.Context, cmd *datasources.UpdateD
 			}
 		}
 
-		// TODO: we will eventually remove this check for moving the resource to it's separate API
-		if s.features != nil && s.features.IsEnabled(ctx, featuremgmt.FlagTeamHttpHeaders) && !cmd.OnlyUpdateLBACRulesFromAPI {
-			s.logger.Debug("Overriding LBAC rules with stored ones",
-				"reason", "update_lbac_rules_from_datasource_api",
-				"action", "use_updateLBACRules_API",
+		// preserve existing lbac rules when updating datasource if we're not updating lbac rules
+		// TODO: Refactor to store lbac rules separate from a datasource
+		if s.features != nil && s.features.IsEnabled(ctx, featuremgmt.FlagTeamHttpHeaders) && !cmd.AllowLBACRuleUpdates {
+			s.logger.Debug("Overriding LBAC rules with stored ones using updateLBACRules API",
```

Comment:
What do you think about extracting the http header preservation code in a separate function, to ease it readability? Something along the line of:

```go
func PreserveTeamHttpHeaders(storedJsonData JsonData , cmdJsonData JsonData) JsonData {
	// If there are no stored data, we should remove the key from the command json data
	if storedJsonData == nil {
		if cmdJsonData != nil {
			cmdJsonData.Del("teamHttpHeaders")
		}
		return cmdJsonData
	}

	previousRules := storedJsonData.Get("teamHttpHeaders").Interface()
	// If there are no previous rules, we should remove the key from the command json data
	if previousRules == nil {
		if cmdJsonData != nil {
			cmdJsonData.Del("teamHttpHeaders")
		}
		return cmdJsonData
	}
	
	if cmdJsonData == nil {
		// It's fine to instantiate a new JsonData here
		// Because it's done in the SQLStore.UpdateDataSource anyway
		cmdJsonData = simplejson.New()
	}
	cmdJsonData.Set("teamHttpHeaders", previousRules)
	return cmdJsonData
}
```